lib/src/string.rs

```diff
@@ -0,0 +1,202 @@
+/// Unless specified otherwise, functions in this API take ownership of a TCString when it is given
+/// as a function argument, and free the string before returning.  Thus the following is valid:
+///
+/// When a TCString appears as a return value or output argument, it is the responsibility of the
+/// caller to free the string.
+pub enum TCString<'a> {
+    CString(CString),
+    CStr(&'a CStr),
+    String(String),
+
+    /// None is the default value for TCString, but this variant is never seen by C code or by Rust
+    /// code outside of this module.
+    None,
+}
+
+impl<'a> Default for TCString<'a> {
+    fn default() -> Self {
+        TCString::None
+    }
+}
+
+impl<'a> TCString<'a> {
+    /// Take a TCString from C as an argument.  C callers generally expect TC functions to take
+    /// ownership of a string, which is what this function does.
+    pub(crate) fn from_arg(tcstring: *mut TCString<'a>) -> Self {
+        debug_assert!(!tcstring.is_null());
+        *(unsafe { Box::from_raw(tcstring) })
+    }
+
+    /// Borrow a TCString from C as an argument.
+    pub(crate) fn from_arg_ref(tcstring: *mut TCString<'a>) -> &'a mut Self {
+        debug_assert!(!tcstring.is_null());
+        unsafe { &mut *tcstring }
+    }
+
+    /// Get a regular Rust &str for this value.
+    pub(crate) fn as_str(&self) -> Result<&str, std::str::Utf8Error> {
+        match self {
+            TCString::CString(cstring) => cstring.as_c_str().to_str(),
+            TCString::CStr(cstr) => cstr.to_str(),
+            TCString::String(string) => Ok(string.as_ref()),
+            TCString::None => unreachable!(),
+        }
+    }
+
+    pub(crate) fn as_bytes(&self) -> &[u8] {
+        match self {
+            TCString::CString(cstring) => cstring.as_bytes(),
+            TCString::CStr(cstr) => cstr.to_bytes(),
+            TCString::String(string) => string.as_bytes(),
+            TCString::None => unreachable!(),
+        }
+    }
+
+    pub(crate) fn to_path_buf(&self) -> PathBuf {
+        // TODO: this is UNIX-specific.
+        let path: &OsStr = OsStr::from_bytes(self.as_bytes());
+        path.to_os_string().into()
+    }
+
+    /// Convert this to a return value for handing off to C.
+    pub(crate) fn return_val(self) -> *mut TCString<'a> {
+        Box::into_raw(Box::new(self))
+    }
+}
+
+impl<'a> From<String> for TCString<'a> {
+    fn from(string: String) -> TCString<'a> {
+        TCString::String(string)
+    }
+}
+
+impl<'a> From<&str> for TCString<'a> {
+    fn from(string: &str) -> TCString<'a> {
+        TCString::String(string.to_string())
+    }
+}
+
+/// Create a new TCString referencing the given C string.  The C string must remain valid until
+/// after the TCString is freed.  It's typically easiest to ensure this by using a static string.
+///
+/// NOTE: this function does _not_ take responsibility for freeing the C string itself.
+/// The underlying string once the TCString has been freed.  Among other times, TCStrings are
```

Comment:
"The underlying string once the TCString has been freed."  sentence seems incomplete